### ISSUE TITLE
single-node: port forward daemon sketch

### ISSUE THREAD

**Issue by philips** [coreos/coreos-kubernetes#444]
This is an idea to make working with the single-node cluster easier. The basic idea would be to have something like `kubectl port-forward` that forwards every node port to localhost. So, for example:

```
$ kubectl run --image quay.io/philips/golang-outyet outyet -l app=outyet
$ kubectl expose deployment outyet --target-port=8080 --port=8080 --type=NodePort
$ socat tcp-listen:8080,reuseaddr,fork tcp:172.17.4.99:$(kubectl get service outyet -o template --template="{{range.spec.ports}}{{.nodePort}}{{end}}")
```

**Comment by aaronlevy** [comment]
I may be missing the motivation, but the nodePort services are available on the host (albeit not via localhost).

For example, in the single-node install the default IP given to the vm is `172.17.4.99`. If you launch the guestbook example using a nodePort type service (lets say it gets assigned port `32000`) it would be available from your host machine via: `curl http://172.17.4.99:32000`

**Comment by oaojesse** [comment]
Seems to me like this would let you treat local containers in development as though they had been deployed into a pod on the cluster.

**Comment by philips** [comment]
@oaojesse another idea would be to use a VPN or something. Then you could access all services and DNS as well.

**Comment by dlorenc** [comment]
The direction I had in mind for minikube was similar to what @aaronlevy pointed out above, in a local single node cluster all nodePort services will be available on the host IP. What if we provided a DNS name via /etc/hosts or similar to that IP? Then it could be as simple as:

curl http://mini.kube:32000

localhost definitely has advantages though. It would be a lot more familiar to most users. It might also come with disadvantages though, we'd have to worry about port collisions more and it would make local multi-cluster harder.

**Comment by oaojesse** [comment]
Thanks for the idea @philips I'll look into that. Hopefully something useful comes out of it.

**Comment by philips** [comment]
Looks like someone has openvpn for k8s: https://github.com/offlinehacker/openvpn-k8s. If someone can try it out and report back that would be great. I will try to get to it in the next few days.

**Comment by philips** [comment]
cc @offlinehacker

**Comment by jessebmiller** [comment]
I'm trying this out today. Hopefully I can get the crypto figured out.

**Comment by oaojesse** [comment]
@philips @offlinehacker

Well I got `openvpn-k8s` running on a cluster on GCP. I was able to figure out the crypto and the vpn all seems to be working. My public IP appears in Mountain View and I can poke services on their cluster IP from my laptop.

I'm not able to use DNS service names yet (despite rebuilding the openvpn-k8s container `from alpine:edge` which fixes a DNS bug).

I expect that I don't understand enough about how kubernetes DNS works yet and I would love a hand if anyone has the time.

**Comment by oaojesse** [comment]
Got DNS working this morning, by copying `/etc/resolv.conf` out of a container in a pod and mounting it into a container running locally while connected to the VPN.

    $ kubectl exec existing-pod cat /etc/resolv.conf > resolv.conf
    $ docker run -it --rm -v `pwd`/resolv.conf:/etc/resolv.conf alpine:edge /bin/ash
    / # nslookup kubernetes
    nslookup: can't resolve '(null)': Name does not resolve
    
    Name:      kubernetes
    Address 1: 10.167.240.1

Everything seems to work great!

**Comment by offlinehacker** [comment]
Hi, i'm glad my openvpn k8s image was helpfull, i haven't used it for some period, as we are now using pritunl, instead of pure openvpn. For dns to work you need to set OVPN_K8S_DNS to ip address of kube-dns service(first entry in /etc/resolv.conf) and then openvpn will push dns configuration. You still need a client that can set dns, like network manager on linux or tunnelblick on osx.

**Comment by offlinehacker** [comment]
Alternative way would be to use something like https://github.com/apenwarr/sshuttle or pptp.

**Comment by oaojesse** [comment]
Yes thank you. It's working great with tunnelblick. I'll play around with getting openvpn to push dns down as that would be nice.
